mlir/lib/Catalyst/Transforms/DetensorizeSCFPass.cpp

```diff
@@ -0,0 +1,354 @@
+        SmallVector<Value> newBlockTransferArgs(newBlock.getArguments().begin(),
+                                                newBlock.getArguments().end());
+        Block &oldBlock = forOp.getRegion().front();
+        rewriter.mergeBlocks(&oldBlock, &newBlock, newBlockTransferArgs);
+
+        // 4. Retensorize arguments at the beginning of the body
+        {
+            OpBuilder::InsertionGuard g(rewriter);
+            rewriter.setInsertionPoint(&newBlock, newBlock.begin());
+            for (std::size_t index : newIterOperandsIndices) {
+                Value value = rewriter.create<tensor::FromElementsOp>(
+                    newForOp->getLoc(),
+                    RankedTensorType::get({}, newForOp.getRegionIterArg(index).getType()),
+                    newForOp.getRegionIterArg(index));
+                rewriter.replaceUsesWithIf(
+                    newForOp.getRegionIterArg(index), value,
+                    [&](OpOperand &op) { return !isa<tensor::FromElementsOp>(op.getOwner()); });
+            }
+        }
+
+        // 5. Extract scalar tensor elements and detensorize terminator
+        {
+            scf::YieldOp clonedYieldOp = cast<scf::YieldOp>(newBlock.getTerminator());
+            SmallVector<Value> newYieldOperands = clonedYieldOp.getOperands();
+            OpBuilder::InsertionGuard g(rewriter);
+            rewriter.setInsertionPoint(clonedYieldOp);
+            for (std::size_t index : newIterOperandsIndices) {
+                newYieldOperands[index] = rewriter.create<tensor::ExtractOp>(
+                    clonedYieldOp->getLoc(), clonedYieldOp->getOperand(index), ValueRange{});
+            }
+            rewriter.create<scf::YieldOp>(newForOp.getLoc(), newYieldOperands);
+            rewriter.eraseOp(clonedYieldOp);
+        }
+
+        // 6. Retensorize results after for loop
+        {
+            OpBuilder::InsertionGuard g(rewriter);
+            rewriter.setInsertionPointAfter(forOp);
+            for (std::size_t index : newIterOperandsIndices) {
+                Value for_result = newForOp->getResult(index);
+                Value value = rewriter.create<tensor::FromElementsOp>(
+                    newForOp->getLoc(), RankedTensorType::get({}, for_result.getType()),
+                    for_result);
+                rewriter.replaceUsesWithIf(forOp->getResult(index), value, [&](OpOperand &op) {
+                    return !isa<tensor::FromElementsOp>(op.getOwner());
+                });
+            }
+        }
+
+        rewriter.replaceOp(forOp, newForOp);
+        return success();
+    }
+};
+
+struct DetensorizeIfOp : public OpRewritePattern<scf::IfOp> {
+    using OpRewritePattern<scf::IfOp>::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(scf::IfOp ifOp, PatternRewriter &rewriter) const override
+    {
+        // Early exit if there are no results that could be replaced.
+        if (ifOp.getNumResults() == 0) {
+            return failure();
+        }
```

Comment:
Very minor, but technically unnecessary since the iterator through the results right below will essentially immediately exit with false.

```suggestion
```